homeassistant/components/zwave_js/__init__.py

```diff
@@ -792,7 +793,11 @@ async def disconnect_client(hass: HomeAssistant, entry: ConfigEntry) -> None:
     for task in platform_setup_tasks:
         task.cancel()
 
-    await asyncio.gather(listen_task, start_client_task, *platform_setup_tasks)
+    tasks = (listen_task, start_client_task, *platform_setup_tasks)
+    await asyncio.gather(*tasks, return_exceptions=True)
```

Comment:
The docs says that a task that is canceled won't stop the other task from being awaited in gather.

https://docs.python.org/3/library/asyncio-task.html#asyncio.gather

Reply:
But whilst the other tasks can still be awaited, they are no longer tracked by the gather, won't be cancelled nor awaited by default, and may fail in the future.

> the first raised exception is immediately propagated to the task that awaits on gather(). Other awaitables in the aws sequence won’t be cancelled and will continue to run.

Reply:
Maybe it's more effective to store the gather task and await it in a loop while suppressing `CancelledError`, until the gather task is done? That way we don't need to loop over all submitted tasks again after they have been awaited with gather.

Reply:
I'm interpreting the docs to say that the gather task will still track the submitted tasks if one of them raises CancelledError, since the gather task is not cancelled.

Reply:
I guess the gather task will raise CancelledError again from a submitted task that was cancelled, though, and that's why we can't await the gather task more than once.

Reply:
Do you agree?

Reply:
Yes. I just ran some tests with a couple of failing task (one after 1s, one after 2s). On first gather, it will wait 1s and then raise the exception:
- task1 is done
- task 2 is still running
- gather is done

Subsequent calls to gather do not wait for the second task, but simply raise again the failure from task1

I couldn't find a way to wait for the "remaining tasks"

Reply:
Maybe you want to do `asyncio.wait` with `ALL_COMPLETED`?

https://docs.python.org/3/library/asyncio-task.html#asyncio.wait

Reply:
Using `await asyncio.wait(tasks)` is the same as `asyncio.gather(*tasks, return_exceptions=True)`

It comes down to error tracking:
- do we want to ignore only CancelledError?, and raise if there was a different error?
- or do we want to ignore all errors?

If we want to ignore all errors, then we can do away with the loop.
If we want to check which errors are not CancelledError, then we need the loop anyway.

Reply:
Let's keep it like this for now. We can improve it with TaskGroup when 3.11 is minimum Python version.

https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup

Reply:
We should probably log any exception with a trace.

We could log `CancelledError` at debug level with a trace as well in case we find a case where cancelation is happening unexpectedly... but that might be do to the cancellation semantics of `asyncio.wait_for` if its used anywhere in the libs which is thankfully fixed in cpython 3.12 https://github.com/python/cpython/pull/98518